File: Tsubame/packages/renderer-dom/src/user-select.ts
import type { ElementKind, UserSelect } from '@torimi/tsubame-renderer-protocol';
import { elementKindDefaultUserSelect } from '@torimi/tsubame-renderer-protocol';

/**
 * 要素の CSS `user-select` 値を解決する（ADR-0108）。DOM Mode はブラウザの
 * ネイティブ選択を使い、`user-select` で範囲を制限する。
 *
 * この意味論は DOM Renderer と Hayate HTML Mode（Rust の `resolve_user_select`）の
 * 唯一の整合基準。共有コーパス `proto/spec/fixtures/user_select_parity.json` が
 * 両者のずれを防ぐ。
 *
 * 解決順序: 明示的 `user-select` → 要素種別の UA 既定 → (none/選択不可)。
 * 選択可能性は CSS と同じくオプトアウト:
 *
 * - `text-input` は常に自身の編集選択を持つので、明示値や種別既定に関わらず `text`。
 * - それ以外は、明示的 `user-select` があればそれ、無ければ種別既定
 *   （`view` / `text` / `scroll-view` = `text`、`image` / `button` = `none`）。
 * - `text` は CSS `text`、`none` は CSS `none`（サブツリーを除外）に対応。
 * - `contains` は選択可能だが包含境界を確立するため CSS `contain` に対応する
 *   （ADR-0108）。`user-select: contain` 対応ブラウザはネイティブ選択を要素内に
 *   閉じ込め、非対応ブラウザは値を無視するがコア側の境界クランプが同じ意味論を
 *   与える（意味論のみの整合）。
 */
export function resolveUserSelect(
  kind: ElementKind,
  explicit: UserSelect | undefined,
): 'text' | 'none' | 'contain' {
  if (kind === 'text-input') return 'text';
  const effective = explicit ?? elementKindDefaultUserSelect(kind);
  if (effective === 'none') return 'none';
  if (effective === 'contains') return 'contain';
  return 'text';
}

File: Hayate/crates/platform/web/src/user_select.rs
//! 要素種別の UA デフォルト + 明示 `user-select` → ブラウザ `user-select` への
//! マッピング（ADR-0108、ADR-0097 を置換）。HTML Mode は `user-select` で
//! 境界づけられたブラウザネイティブ選択を使う。Tsubame DOM Renderer
//! （`resolveUserSelect`）とのパリティ契約の Rust 側で、単一ソース
//! `proto/spec/fixtures/user_select_parity.json` が両側を固定する（ADR-0070）。

use hayate_core::{ElementKind, UserSelectValue};

/// 要素の CSS `user-select` 値を解決する。
///
/// 解決順序（ADR-0108）: 明示 `user-select` → 要素種別の UA デフォルト →
/// （none/選択不可）。選択可能性は CSS と同様にオプトアウト。
///
/// - text-input は常に編集選択を持つため、明示値や種別デフォルトに関わらず `text`。
/// - それ以外の実効値は、明示 `user-select` があればそれ、無ければ種別デフォルト
///   （`view` / `text` / `scroll-view` = `text`、`image` / `button` = `none`）。
/// - `text` は CSS `text`、`none` は CSS `none` へ。後者はサブツリーを除外する。
/// - `contains` は選択可能だが包含境界を確立するため CSS `contain` へ写す
///   （ADR-0108）。`user-select: contain` をサポートするブラウザはネイティブ選択を
///   要素内に閉じ込め、非サポートのブラウザはそれを無視するが、core 側の境界
///   クランプが同じセマンティクスを供給する（セマンティクスのみのパリティ）。
pub fn resolve_user_select(kind: ElementKind, explicit: Option<UserSelectValue>) -> &'static str {
    if matches!(kind, ElementKind::TextInput) {
        return "text";
    }
    let effective = explicit.unwrap_or_else(|| kind.default_user_select());
    match effective {
        UserSelectValue::Text => "text",
        UserSelectValue::Contains => "contain",
        UserSelectValue::None => "none",
    }
}
